relayer/chain/ethereum/syncer/syncer.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package syncer
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+type latestBlockInfo struct {
+	sync.Mutex
+	fetchFinalizedDone bool
+	height             uint64
+}
+
+// Syncer retrieves headers starting at a given initial height up to the latest.
+// Headers are sent to the channel `headers` in order. If the initial height is
+// old, (finalized) headers in the canonical chain will be forwarded in quick succession
+// until we catch up with the unfinalized headers. From that point onwards, headers
+// on all forks are forwarded. A header is considered final if it has at least
+// `descendantsUntilFinal` descendants.
+type Syncer struct {
+	descendantsUntilFinal uint64
+	headerCache           HeaderCache
+	headers               chan<- *gethTypes.Header
+	loader                HeaderLoader
+	log                   *logrus.Entry
+}
+
+func NewSyncer(descendantsUntilFinal uint64, loader HeaderLoader, headers chan<- *gethTypes.Header, log *logrus.Entry) *Syncer {
+	return &Syncer{
+		descendantsUntilFinal: descendantsUntilFinal,
+		headerCache:           *NewHeaderCache(descendantsUntilFinal + 1),
+		headers:               headers,
+		loader:                loader,
+		log:                   log,
+	}
+}
+
+func (s *Syncer) StartSync(ctx context.Context, eg *errgroup.Group, initBlockHeight uint64) error {
+	latestHeader, err := s.loader.HeaderByNumber(ctx, nil)
+	if err != nil {
+		s.log.WithError(err).Error("Failed to retrieve latest header")
+		return err
+	}
+
+	lbi := &latestBlockInfo{
+		fetchFinalizedDone: false,
+		height:             latestHeader.Number.Uint64(),
+	}
+
+	eg.Go(func() error {
+		return s.fetchFinalizedHeaders(ctx, initBlockHeight, lbi)
+	})
+	eg.Go(func() error {
+		return s.pollNewHeaders(ctx, lbi)
+	})
+
+	return nil
+}
+
+func (s *Syncer) fetchFinalizedHeaders(ctx context.Context, initBlockHeight uint64, lbi *latestBlockInfo) error {
+	syncedUpUntil := initBlockHeight
+
+	for {
+		lbi.Lock()
+		latestFinalizedHeight := lbi.height - s.descendantsUntilFinal
+		if syncedUpUntil >= latestFinalizedHeight {
+			// Signals to pollNewHeaders that new headers can be forwarded now
+			lbi.fetchFinalizedDone = true
+			lbi.Unlock()
+
+			s.log.WithField("blockNumber", syncedUpUntil).Debug("Done retrieving finalized headers")
+
+			break
+		}
+		lbi.Unlock()
+
+		header, err := s.loader.HeaderByNumber(ctx, new(big.Int).SetUint64(syncedUpUntil+1))
```

Comment:
Fetching by height here only gives you a chain if the node we are connected to has the same definition of finality. If they are more conservative (they require more confimrations), the node could give us headers from another fork at some point.

I think we should check whether the headers connect in this group, and reset the loop if we find out that the fork changed. Alternatively, we could just go backwards from finalized, and query headers by hash.

What do you think?

Reply:
> Fetching by height here only gives you a chain if the node we are connected to has the same definition of finality.

This is a fair point.

> I think we should check whether the headers connect in this group, and reset the loop if we find out that the fork changed. Alternatively, we could just go backwards from finalized, and query headers by hash.

Sounds sensible to put a sanity check that headers connect here and rewinding if they don't - but how far does one rewind? If the `descendantsUntilFinal` value of 35 isn't sufficient to prevent this from happening, what value is? The alternative (fetching backwards) doesn't scale because it forces serial fetching. I'd like to give us the option of concurrent fetching + batching in future. Thoughts?

Reply:
Good points. 

I guess it's fine to not have any checks for now, since a failure would just mean that the light clients will stop proceeding, which should be monitored anyways. A simple restart of the relayer should suffice to fix the issue. 

It would be a bit prettier if we actually check whether the headers connect, and just fail the relayer if they don't. 
I'd be fine though to just add a TODO explaining those potential issues for now.

Reply:
Failing loudly in the relayer is a good idea. I'll also add a comment